### PR TITLE
Finalize handling of IObjectOrCollectionInitializerOperation in control flow graph.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-infrastructure It looks like CI jobs are not making progress.

**Comment by AlekseyTs** [comment]
@dotnet-bot Test this please

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Is it likely that we are missing on handling more such operations kinds in this switch statement? If so, should we consider changing the default clause to assert and return instead of throwing?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Is it likely that we are missing on handling more such operations kinds in this switch statement? If so, should we consider changing the default clause to assert and return instead of throwing?

This is likely what we will end up doing, even without an assert. However, we should look at the scenarios to make sure we are not missing something important

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > We are dealing with erroneous object creation. All children, but the last one are arguments for the constructor, [](start = 18, length = 113)

I feel uncomfortable with this reverse engineering (inferring from the shape of the invalid operation how it was produced and what it means). 
Is there a reason we can't produce an object creation (with some bits missing) even when overload resolution failed?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > I feel uncomfortable with this reverse engineering (inferring from the shape of the invalid operation how it was produced and what it means). 
Is there a reason we can't produce an object creation (with some bits missing) even when overload resolution failed?

This is definitely an error case (an invalid operation), we just need to have reasonable recovery story. ObjectOrCollectionInitializer cannot appear in any context other than and attempt to create an initialize an instance. Ignoring any possible bugs in the compiler, we are pretty safe to make that assumption. We are not creating object creation in the case of overload resolution because we cannot provide symbol for constructor and reliable mapping from arguments to parameters, default values and paramarrays.

---
In reply to: [192865269](https://github.com/dotnet/roslyn/pull/27393#discussion_r192865269) [](ancestors = 192865269)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1).

**Review by jcouv — APPROVED** [review]
LGTM following discussion. Thanks

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need second review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: If you have to fix up merge conflicts, consider reverting this change I have the TypeParameterObjectCreation pr ready to go as soon as tests pass on [https://github.com/dotnet/roslyn/pull/27308](https://github.com/dotnet/roslyn/pull/27308).

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > If you have to fix up merge conflicts, consider reverting this change I have the TypeParameterObjectCreation pr ready to go as soon as tests pass on [https://github.com/dotnet/roslyn/pull/27308](https://github.com/dotnet/roslyn/pull/27308).

I think the test hook will not be "green" without this change. We want to remain "green" all the time

---
In reply to: [192874642](https://github.com/dotnet/roslyn/pull/27393#discussion_r192874642) [](ancestors = 192874642)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Consider an additional test with dataflow in the initializer instead of the constructor.

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 1)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: > Consider an additional test with dataflow in the initializer instead of the constructor.

Sure, I'll a test like that in one of the next PRs

---
In reply to: [192875722](https://github.com/dotnet/roslyn/pull/27393#discussion_r192875722) [](ancestors = 192875722)

**Pull request merged**
